internal/backend/runtime/omni/controllers/omni/internal/task/machine/poll.go

```diff
@@ -340,3 +346,48 @@ func pollExtensions(ctx context.Context, c *client.Client, info *Info) error {
 
 	return nil
 }
+
+func pollMaintenanceConfig(ctx context.Context, c *client.Client, info *Info) error {
+	if !info.MaintenanceMode {
+		return nil
+	}
+
+	conf, err := safe.StateGetByID[*config.MachineConfig](ctx, c.COSI, config.V1Alpha1ID)
+	if err != nil {
+		// if we successfully probed the config and found out that it is not set, we mark it as ready by setting it to a non-nil value by empty config.
+		if state.IsNotFoundError(err) {
+			info.MaintenanceConfig = &specs.MachineStatusSpec_MaintenanceConfig{}
+
+			return nil
+		}
+
+		// this resource might not be authorized to be read over siderolink in the maintenance mode on Talos versions < 1.6.5.
+		// in that case, if it was never set previously, we mark it as ready by setting it to a non-nil value with empty config.
+		if code := status.Code(err); code == codes.PermissionDenied {
+			if info.MaintenanceConfig == nil {
+				info.MaintenanceConfig = &specs.MachineStatusSpec_MaintenanceConfig{}
+			}
+
+			return nil
+		}
+
+		return fmt.Errorf("failed to get maintenance config: %w", err)
+	}
+
+	for _, document := range conf.Container().Documents() {
+		if document.Kind() == v1alpha1.Version {
+			return fmt.Errorf("unexpected v1alpha1.Config document in maintenance config: %w", err)
+		}
+	}
+
+	configBytes, err := conf.Container().Bytes()
+	if err != nil {
+		return fmt.Errorf("failed to read maintenance config bytes: %w", err)
+	}
+
+	info.MaintenanceConfig = &specs.MachineStatusSpec_MaintenanceConfig{
```

Comment:
if we're looping over multiple documents, aren't we setting the last one only?

Reply:
The loop is to make sure that there is no `v1alpha1` document in the config. Then, we get the whole config we found and set it on the `MachineStatus`. This is an example of what we could find:
```
apiVersion: v1alpha1
kind: SideroLinkConfig
apiUrl: grpc://omni.omni-local.utkuozdemir.org:8090?jointoken=w7uVuW3zbVKIYQuzEcyetAHeYMeo5q2L9RvkAVfCfSCD
---
apiVersion: v1alpha1
kind: EventSinkConfig
endpoint: "[fdae:41e4:649b:9303::1]:8090"
---
apiVersion: v1alpha1
kind: KmsgLogConfig
name: omni-kmsg
url: "tcp://[fdae:41e4:649b:9303::1]:8092"
```

The loop is to fail fast if we see a `version: v1alpha1` doc among them (which shouldn't actually happen).

Reply:
oh, nevermind, I missed the loop closure :sweat_smile: